Generic model builder and inference engine

wyessen opened the pull request:
Quite a bit to chew on in this PR. The main features are:

- Generic model builder (x-model-builder)
- Generic inference engine (x-infer)
- Documentation for the generic model builder and the generic inference engine

ayermolo commented:
Please make sure it doesn't break windows build.  I saw dlopen which I believe is linux only.

Also maybe refactor in to more manageable commits to make review easier. :)

wyessen commented:
> Please make sure it doesn't break windows build. I saw dlopen which I believe is linux only.
> 
> Also maybe refactor in to more manageable commits to make review easier. :)

You're right, the inference engine is not supported on Windows. It isn't just calls to `dlopen` -- there are many more dependencies that assume UNIX-like platforms. This is made explicit in the accompanying documentation. This is also the reason why the inference engine is not built as part of Glow (but CMake build system is provided for the end user to build independently of Glow). 

Porting to windows would involve a nontrivial amount work. The main motivation for the existence of the engine is to build for various UNIX-like platforms, including the ability to cross-compile for those embedded systems that support UNIX and POSIX. So extension to whatever unsupported platforms would be the responsibility of the developer targeting those platforms. In other words, it is not meant to be generic, but serve as an extensive (and functional) example.

wyessen commented:
@jackm321 Please let me know if you need anything from me that could help you with the review.

likethesky commented on inference_engines/x-inference-engines/x_perf_monitor.h:
```diff
@@ -0,0 +1,50 @@
+/** Copyright 2019 Xperi Corporation.
```
Yes, by signing the CLA, all contributors are agreeing to that and although of course anything you write is by definition copyright by the author, the controlling agreement is the CLA and the license that Glow is provided under. So, yes, in order to avoid confusion, all copyright headers should match what we are using in the rest of Glow. Thanks Jack for making sure that's the case--and thank you very kindly @wyessen for your contribution(s)!

wyessen replied:
Thanks @jackm321 for the review! I'll be getting to the technical points you made shortly.

@likethesky On legal matters: our contribution is of course provided under the same license as Glow, and we did sign the CLA. The only thing that is different in the headers is the "Copyright 2019 Xperi Corporation" rather than "Copyright ... Facebook...". 

So my question is whether we should change our header only to avoid confusion, or there is legal pressure? Just trying to gather enough info before I kick this over to the relevant parties on our end. Thanks!

likethesky replied:
I checked w/legal before replying the first time. So yes, some legal oversight has occurred and that's their rec. Of course, the Apache 2.0 license allows anyone to copy/use any of this code at any time, so you guys are protected. Also, I probably misspoke or wasn't totally clear in my statement above, when you "write" anything, it's copyright you, but once you contribute it under our CLA, the copyright is granted to Facebook. See under "2. Grant of Copyright License" here: https://code.facebook.com/cla/

Hope that helps. So yes, needs to be (c) Facebook, please use the standard copyright used throughout Glow (and I believe every other of the thousand+ OSS projects we have). Cheers!

likethesky commented on docs/InferenceEngines.md:
```diff
@@ -0,0 +1,521 @@
+
+### A Note on the Initial Contribution
+
+The initial contribution of the `x-inference-engine` (as well as the corresponding documentation) was made as part of the open source contribution initiative by the XPERI Corporation (xperi.com). 
```
Okay; a bit unusual, but allowable. Doesn't change the grant of copyright, or the Glow license in any way. I'd generally discourage these notations, just because they slow us down a bit to confirm wording w/legal, but if they want to make sure their name is in here somewhere, counsel says okay.

wyessen replied:
> Okay; a bit unusual, but allowable. Doesn't change the grant of copyright, or the Glow license in any way. I'd generally discourage these notations, just because they slow us down a bit to confirm wording w/legal, but if they want to make sure their name is in here somewhere, counsel says okay.

@likethesky Understood. Our legal team is OK with us changing the headers to what Glow uses everywhere else. One question, though: are we allowed to identify ourselves somewhere following the source code legal headers? Something like the following, perhaps:

`/* Contributed by Xperi Corporation on [date] */`

So perhaps the entire header could look like this:

```
/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Contributed by Xperi Corporation on [date] 
 */

It doesn't really matter where we identify ourselves, but it feels that somewhere in the header is the most natural place. Thanks!

jackm321 commented:
> We are potentially targeting embedded hardware for which there might not exist a decent C++ compiler. This decision was also (only slightly) reinforced by the choice of Glow to produce C bundles. So, C was really chosen as the smallest common denominator.

Ok I see, thank you for explaining.


> Certainly, some rudimentary "sanity" tests can be provided. Perhaps we should make this a separate ticket? It will take me a while to do this, as I'm currently low on bandwidth.

Sounds good.

wyessen commented:
@jackm321 I believe I've addressed all your concerns in this latest commit. Please have a look. Thanks!

As agreed, tests will be provided as a separate PR.

wyessen commented:
> @jackm321 I believe I've addressed all your concerns in this latest commit. Please have a look. Thanks!
> 
> As agreed, tests will be provided as a separate PR.

@jackm321 A very gentle bump :). I have some bandwidth to start working on more additions which will be based on this, so I'd like to know if this is fine. Thanks for all your time!

wyessen commented:
> @wyessen sorry, I was out on pto last week, I've taken another look and it looks good to me!
> The only things I'd request to change are
> 
> 1. also match style for variable and function names for example `initConstantWeights` instead of `init_constant_weights`
> 2. Add `inference_engines/` to format.sh

Hi Jack,

Great, thanks! Ok, will make those changes.

wyessen commented:
> > @wyessen sorry, I was out on pto last week, I've taken another look and it looks good to me!
> > The only things I'd request to change are
> > 
> > 1. also match style for variable and function names for example `initConstantWeights` instead of `init_constant_weights`
> > 2. Add `inference_engines/` to format.sh
> 
> Hi Jack,
> 
> Great, thanks! Ok, will make those changes.

Ok, all done.
